cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/HttpRequestHandler.java

```diff
@@ -157,6 +158,14 @@ private HttpRequest applyProxyRules(HttpRequest request) {
   public void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e)  {
     Throwable cause = e.getCause();
 
+    // avoid handling exception more than once from a handler, to avoid a possible infinite recursion
+    if (exceptionHandled) {
+      LOG.error("Not handling exception due to already having handled an exception in Request Handler {}",
+                ctx.getChannel(), cause);
+      return;
+    }
+    exceptionHandled = true;
```

Comment:
Should do it slightly differently. In case the `LOG` statement throw exception, it can still runs into infinite recursive call. It's better be using a `AtomicBoolean` to help:

```java
private final AtomicBoolean exceptionHandled = new AtomicBoolean();
...
if (!exceptionHandled.compareAndSet(false, true)) {
  LOG.error("Ignoring exception");
  return;
}
```